chore(rfcs): Remap support for template strings

FungusHumungus opened the pull request:
Ref #4904 

[Readable version](https://github.com/timberio/vector/blob/4909_remap-template/rfcs/2020-11-06-4909-remap-template-fields.md)

Note, this RFC does suggest the simplest solution - just remove support for field lists in template fields. I'm not necessarily recommending that, I just figure it's best to start simple and work up from there!


Signed-off-by: Stephen Wakely <fungus.humungus@gmail.com>

JeanMertz commented on rfcs/2020-11-06-4909-remap-template-fields.md:
```diff
@@ -0,0 +1,142 @@
+
+As there are a number of complications with working out which fields are used
+in the output, we could only support this feature if the old style template
+syntax is used. If Remap is used, the user won't have the option to remove these
+fields from the message in the Loki sink.
```
I actually have on my list a change to the `del` function for it to return the value of the last field it deletes (this is part of an effort to get rid of the `None` return value in Remap, which brings with it a whole bunch of tricky to catch issues).

Given that, it would be possible for someone to write:

```
labels.key = "{{ del(.foo) }}-{{ del(.buz) }}"
```

To achieve the same result, using Remap.

JeanMertz replied:
Even without that change, it would still be possible (buy way less ergonomic) to achieve the same result with the current Remap feature-set:

```lua
labels.key = """{{
$foo = .foo
del(.foo)

$buz = .buz
del(.buz)

$foo + "-" + $buz
}}
"""
```

JeanMertz commented on rfcs/2020-11-06-4909-remap-template-fields.md:
```diff
@@ -0,0 +1,142 @@
+Remap is parsing the script, it could keep track of all fields that are being
+used in the script.
+
+This script, `if .foo { .bar } else { .baz }`, would result in all three
+fields being returned - `.foo`, `.bar` and `.baz`, and subsequently removed
+from the message sent to Loki.
```
This seems way too surprising, and to me, disqualifies this alternative.

binarylogic replied:
Agree. Let's leave Loki's special case out of this RFC and handle it separately. I am not a fan of this, and it's likely we'll handle field deletion separately in an explicit manner. In the interim, users can continue to use the old templating until we solve this.

JeanMertz commented on rfcs/2020-11-06-4909-remap-template-fields.md:
```diff
@@ -0,0 +1,142 @@
+condition and those used it the result, so only `.bar` or `.baz` could be
+returned.
+
+There are likely to be a number of edge cases that would need to be thought
+through if we took this approach. For example, if a field is used in a function
+but it's value is only used to influence the result - should it be included?
```
Yeah, this seems like a no-go, again because there are way too many edge-cases and the added cognitive complexity it adds to the language.

JeanMertz reviewed (changes requested):
I left a few comments, showing that it's actually possible (barely, currently, but with some modifications, more ergonomically) for remap-based templates to delete fields while they are being processed.

The one thing this relies on, is for Remap to have mutable access to the event. That's doable, but perhaps surprising in the context of templating? With the work in #4744, we could define which configuration fields support mutable access to the event and which don't, but it's definitely something we need to consider carefully.

I'm not a fan of the proposed alternatives, given the added (cognitive) complexity.

From reading this RFC, it also isn't entirely clear to me what this RFC is currently proposing.

It seems to me that the first step towards a solution for this is to support Remap for templating _in general_, regardless of the specific issue with deleting fields as is the case in the `loki` sink.

I don't see anything written about the syntax used to use the Remap language in templates without breaking backward compatibility, or a discussion on which Remap features are available (e.g. mutating events, which functions do we expose, etc).

It feels as if the RFC skips several steps to end up discussing how to handle deleting fields, whereas the steps that come before this aren't as clear to me yet.

Is it worth discussing that in this RFC, or should we have a separate RFC first to discuss how to add Remap as an option for string templates?

JeanMertz reviewed (changes requested):
The RFC itself in terms of details is going in the right direction, but I'm not convinced of the proposed solution yet (but also don't have one myself that isn't flawed in some way).

Curious what other people think.

Also, it seems we should just name this RFC "Support remap in string templates", as that's what's being discussed here (field lists and/or deleting those lazily is part of the proposal, but does not encapsulate everything in it).

jszwedko commented on rfcs/2020-11-06-4909-remap-template-fields.md:
```diff
@@ -0,0 +1,191 @@
+
+## Outstanding Questions
+
+- How important is it for the Loki sink to remove fields used in the template
```
I'd propose deprecating `remove_label_fields`; allowing its use only with non-remap templates.

I think, going forward, a better way to handle options like `remove_label_fields` is to let users use `remap` in the component itself. That is, they'd do something like:

```toml
[sinks.loki]
  type = "loki"
  inputs = ["json"]
  endpoint = "http://localhost:3100"
  remove_label_fields = true
  labels.key = "{{{ .foo }}}-{{{ .buzz }}}"
  remap = """del(.foo, .buzz)"""
```

JeanMertz replied:
I agree with this. This is a clearly superior solution 👍

I'd probably vote to have two config fields available in _all_ components:

* `remap.before`
* `remap.after`

Both take a Remap script, and run either before or after the other logic of the component finished.

binarylogic replied:
> I'd probably vote to have two config fields available in all components:

Curious what @lukesteensen thinks about this. We've had similar discussions for control flow:

```toml
[transforms.sampler]
  type = "sampler"
  
  [[if]]
    condition = ".status == 200"
    # regular component options here.

  [[if]]
    condition = ".status == 500"
    # regular component options here.
```

jszwedko reviewed (commented):
Nice write-up!

I think I lean towards `{{{ }}}` to indicate remap style templates to let people opt into it and avoid any ambiguity. As Jean notes, we can deprecate the simple `{{ }}` templates and eventually let `{{ }}` indicate a remap template too.

FungusHumungus commented:
Ok, I have now updated this to specify:

- `{{{..}}}` syntax for Remap template fields, `{{.}}` remains for old style template fields. We can unambiguously distinguish between the two.
- Added `remap.before` and `remap.after` fields to the config for any processing that may need doing, including removing any fields from the event. This allows us to deprecate the `remove_label_fields`.

I'm feeling good about this now. I think we have a pretty clean way forward.

JeanMertz commented on rfcs/2020-11-06-4909-remap-template-fields.md:
```diff
@@ -0,0 +1,211 @@
+  type = "loki"
+  inputs = ["json"]
+  endpoint = "http://localhost:3100"
+  remove_label_fields = true
```
Should we just drop this config straight away?

Even if people still use the original `{{ ... }}` templating, they still can replace `remove_label_fields` with `remap.after`. This change would be a breaking change, but one we're going to make at some point regardless, so might as well get it out of the way.

JeanMertz commented on rfcs/2020-11-06-4909-remap-template-fields.md:
```diff
@@ -0,0 +1,211 @@
+
+### Add `remap.before` and `remap.after` fields
+
+To replicate this functionality we will add `remap.before` and `remap.after`
```
So, I've thought about `remap.before` and `remap.after` a bit more, and looking at this example, it's not so clear what `remap.after` means in _this_ context, and how it would relate to the same field in _other_ components.

For the `loki` sink, to have `remap.after` be useful, it would have to fire _before_ the data is sent to Loki, but _after_ the `labels.key` is applied.

However, for transforms or sources, `remap.after` would just run "right before the transform/source is done with the event".

So, I guess `remap.after` is described as

> Runs _before_ the end of the component. For **sources** and **transforms** this is _before_ the event is passed to the next component, for **sinks** this is _before_ the event is delivered by Vector to the configured sink.

For `remap.before`, it would be similar, except that sources/sinks are swapped:

> Runs right _after_ the start of the component. For **transforms** and **sinks** this is _before_ the event is processed by the given component, for **sources** this is _after_ the event is received by Vector, but before the source has processed the event.

It requires a bit of explaining, but it does provide the most amount of flexibility.

jszwedko replied:
:+1: I think this is the most easily understandable way to model it.

lukesteensen replied:
It's not entirely clear to me why we need two variants. For sources, it really only makes sense to modify the event after we've received it, and likewise, for sinks, we can only change it before we send it.

Transforms do have a before and after, but it seems confusing to wrap one transform with what's effectively just another transform. Do we have a good example of a transform that'd benefit from being able to have pre- or post-processing with remap without just chaining a remap transform? It seems to me that a single `remap` option on sources and sinks should cover the vast majority of use cases.

Coming back to templating specifically, I'm not sure it makes sense to encourage users to use multiple different remap executions (one generic pre-sink and one per templated field) and try to explain their ordering. Why not just have one remap execution that happens first and leave the templating as simple field access?

Initially, we can just leave the options to remove templated fields as-is, but eventually, with the introduction of better event metadata, we could give remap access to that metadata and allow users to remove fields from the normal event data while still having access via metadata for templated fields.

I understand the desire to use remap for literally everything, but there is a reason that most templating solutions use a dedicated language.

JeanMertz replied:
Agreed, there are considerations to be made here. We agreed that this RFC is not the place to introduce `remap.before/after` or any other variant.

I'm sure whoever picks up that RFC work can expand on the use-case for before/after (or lack thereof).

JeanMertz reviewed (changes requested):
Nicely done @FungusHumungus.

The `remap.before` and `remap.after` global config fields has a fairly significant impact on Vector itself.

I would almost advocate we split that part out of this RFC into a new one, that describes all the intrinsic details and then ask more people to weigh in on that proposal (especially because it's also closely tied to using conditional logic in the Vector config itself, as @binarylogic mentioned).

Then, once that RFC is approved and implemented, we can just PR an update to the `loki` sink to remove `remove_label_fields` favouring `remap.after`, and then finally we can create a separate (small) RFC to propose the new remap-inspired template fields using `{{{ ... }}}` (probably just updating this one, and removing all references to `loki` and the `remap.before` stuff).

It's maybe a bit more involved than just getting this RFC in, but I'm not comfortable enough that we've covered our bases with the impact that `remap.before/after` can have on Vector as a whole.

Having said that, if @binarylogic considers this unneeded, then I'm happy to see this RFC merged, as there's nothing blocking any more AFAICT.

binarylogic commented:
(note, I talked with Stephen on Discord about this)

Thanks for putting this together @FungusHumungus! Given the various issues (deleting `loki` sink labels, using private data in templates), we are leaning towards doing nothing and reapproaching this later. There are a few reasons for this:

1. We are working on adding event metadata (#5802) as part of the various e2e work. We will allow the reading of metadata in templates which solves the issue of using event data in templates that should not be encoded.
2. If a user needs to use complex Remap functions to generate specific configuration values then they should do this in a `remap` transform before the sink. Not only does this simplify observability of Vector and readability of Vector configs, but it also simplifies templating. The user can just call the field (`{{ special_field_value }}`) directly. It is possible that we will allow users to set custom metadata values (see point 1) which would resolve the need to exclude these values when encoding the event.

Given these 2 points, let's close this and see where that takes us. It's possible this is enough.
